File: telegram-service/src/main/java/de/nak/home_assistant/commands/Input.java
package de.nak.home_assistant.commands;

import com.pengrad.telegrambot.TelegramBot;
import com.pengrad.telegrambot.model.Message;
import com.pengrad.telegrambot.model.request.InlineKeyboardButton;
import com.pengrad.telegrambot.model.request.InlineKeyboardMarkup;
import com.pengrad.telegrambot.request.DeleteMessage;
import com.pengrad.telegrambot.request.GetChat;
import com.pengrad.telegrambot.request.SendMessage;
import com.pengrad.telegrambot.request.UnpinChatMessage;
import com.pengrad.telegrambot.response.GetChatResponse;
import de.nak.home_assistant.actions.DeviceAction;
import de.nak.home_assistant.models.telegram.CallbackData;

public class Input {

    private final TelegramBot bot;

    public Input(TelegramBot bot) {
        this.bot = bot;
    }

    private void sendError(long chatId) {
        SendMessage msg = new SendMessage(chatId, "Ich weiß nicht was ich tun soll");
        bot.execute(msg);
    }

    public void replyToReply(Message reply) {
        long chatId = reply.chat().id();

        if (reply.replyToMessage() == null) {
            sendError(chatId);
            return;
        }

        // Delete input hint
        bot.execute(new DeleteMessage(chatId, reply.replyToMessage().messageId()));

        GetChatResponse chatResponse = bot.execute(new GetChat(chatId));
        Message originalMessage = chatResponse.chat().pinnedMessage();

        if (originalMessage == null) {
            sendError(chatId);
            return;
        }

        bot.execute(new UnpinChatMessage(chatId).messageId(originalMessage.messageId()));

        InlineKeyboardMarkup markup = originalMessage.replyMarkup();
        if (markup == null
                || markup.inlineKeyboard() == null
                || markup.inlineKeyboard()[0].length == 0
                || markup.inlineKeyboard()[0][0] == null) {

            sendError(chatId);
            return;
        }

        InlineKeyboardButton button = markup.inlineKeyboard()[0][0];
        String cbString = button.callbackData();
        CallbackData cbData = new CallbackData().fromJson(cbString);

        long userId = reply.from().id();
        DeviceAction.generateMessages(userId, chatId, cbData, originalMessage, reply.text())
                .forEach(bot::execute);
    }
}

File: telegram-service/src/commands/input.ts
import { Message } from "node-telegram-bot-api";
import { sendDeviceAction } from "../actions/device-action";
import { bot } from "../bot";
import { i18n } from "../i18n";
import { CallbackData } from "../telegram/callback-data";

export const replyToReply = async ({
  reply_to_message,
  chat,
  from,
  text,
}: Message) => {
  if (!reply_to_message) {
    return;
  }
  const translations = i18n();

  await bot.deleteMessage(chat.id, reply_to_message.message_id.toString());

  const { pinned_message } = await bot.getChat(chat.id);
  if (!pinned_message) {
    return bot.sendMessage(chat.id, translations.input.pinnedMessageError);
  }

  await bot.unpinAllChatMessages(chat.id);

  const markup = pinned_message.reply_markup;
  if (
    !markup ||
    markup.inline_keyboard.length == 0 ||
    markup.inline_keyboard[0].length == 0 ||
    !markup.inline_keyboard[0][0]
  ) {
    return bot.sendMessage(chat.id, translations.input.markupError);
  }

  const button = markup.inline_keyboard[0][0];
  const cbData: CallbackData = JSON.parse(button.callback_data);
  return sendDeviceAction(from.id, chat.id, cbData, pinned_message, text);
};
